### ISSUE TITLE
'listen' in config.json can be counter-intuitive

### ISSUE THREAD

**Issue by pepedocs** [cgrates/cgrates#151]
Hi,

I just noticed that the config key in config.json called 'listen' can be counter-intuitive to some users, because it can mean 2 things;

1.  Listen to these addresses (ingress)
2.  Send requests to these addresses (egress)

It may not be an issue when installing 1 cgrates in one instance/server, but when cgrates components are separated in different servers this might cause confusion.  For instance, I'd like my cgrates servers to use a http://non-local:2080 endpoint, but in the config it isn't intuitive enough to infer that I just simply need to modify the 'listen' config, because listen again could mean #1.  

Sorry if this sounds like nitpicking but adding a config for endpoints I think can be more intuitive than reusing 'listen'.  

Additional config may look like this;

"endpoints": {
    "http://cgrates.apier.server.domain.com:2080",
}

Cheers,
Jose

**Comment by danbogos** [comment]
Hi Jose,

This one might be misunderstanding from your side.

The listen configuration will tell on which  ports will CGRateS 
listen/bind when started. Since the ports are shared by more subsystems 
it is more like a generic configuration than stating which services are 
started or not. Services will be started in configs bellow listen part.
Now, we have used "listen" since that is a more widely accepted 
directive (eg: all kamailio/opensips/freeswitch/apache, etc using same 
listen when binding to ports).

Hope this one makes sense to you.

DanB

**Comment by pepedocs** [comment]
Hi Dan,

Please read my explanation below, this is really more of an improvement
than issue.

On Fri, Aug 21, 2015 at 5:40 PM, Dan Christian Bogos <
notifications@github.com> wrote:

> Hi Jose,
>
> This one might be misunderstanding from your side.
>
> The listen configuration will tell on which ports will CGRateS
> listen/bind when started. Since the ports are shared by more subsystems
> it is more like a generic configuration than stating which services are
> started or not. Services will be started in configs bellow listen part.
> Now, we have used "listen" since that is a more widely accepted
> directive (eg: all kamailio/opensips/freeswitch/apache, etc using same
> listen when binding to ports).
>

I completely understand all of these, in fact the reason behind my
confusion is that the 'listen' is conventionally used for a process to bind
on a socket (port and ip address) and that is my perspective as I freshly
come from a cloud systems engineering background.

Maybe I have not explained my first post very well, so I would like to ask
a question instead.

If I have a cgrates instance in server A, with a 'listen'  config.
declaring the following;

'listen': 'http://serverB:2080' ,

Question:  Since the serverB:2080 address is not local ( non-localhost
rfc6761), will this tell server A to send http requests to server B ?

Unless my testing is flawed, I conclude that in this case the 'listen'
config is used is used as an 'endpoint' rather than a socket 'bind'.


>
> Hope this one makes sense to you.
>
> DanB
>
> On 21.08.2015 05:43, Jose Cueto wrote:
> >
> > Hi,
> >
> > I just noticed that the config key in config.json called 'listen' can
> > be counter-intuitive to some users, because it can mean 2 things;
> >
> > 1. Listen to these addresses (ingress)
> > 2. Send requests to these addresses (egress)
> >
> > It may not be an issue when installing 1 cgrates in one
> > instance/server, but when cgrates components are separated in
> > different servers this might cause confusion. For instance, I'd like
> > my cgrates servers to use a http://:2080 endpoint, but in the config
> > it isn't intuitive enough to infer that I just simply need to modify
> > the 'listen' config, because listen again could mean #1
> > <https://github.com/cgrates/cgrates/issues/1>.
> >
> > Sorry if this sounds like nitpicking but adding a config for endpoints
> > I think can be more intuitive than reusing 'listen'.
> >
> > Cheers,
> > Jose
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/cgrates/cgrates/issues/151>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cgrates/cgrates/issues/151#issuecomment-133316497>.
>



-- 
To stop learning is like to stop loving.

**Comment by danbogos** [comment]
Hi Jose,

Listen is the socket bind, so not remote connect. It will define where 
local service (as endpoint) will listen for remote to connect.

Does that make sense?

DanB

**Comment by danbogos** [comment]
Hi Jose,

Please let us know whether all clear for now so we can close this issue.

DanB

**Comment by pepedocs** [comment]
Hi Dan,

Thanks for your reply, and sorry for posting an issue so soon.  What made me think about remote connection or #2 above is that I saw http packets coming out exactly from that ip:port which I can only remember cgrates to be configured to bind to.  I found out it was another internal process in our cloud provider which was the one generating those packets.

So I conclude cgrates isn't doing anything besides listening with the 'listen' config, and that it is doing perfectly what it is supposed to.

Cheers
